### PR TITLE
refactor(agg): decouple aggregate function and state

### LINKED ISSUE #9900: agg: reduce agg state size and memory allocation
### Is your feature request related to a problem? Please describe.

In #9864 we found that the aggregator state size was incredibly high.

A simple `sum(int64)` state consumes up to 96 bytes memory and has 3 mallocs:
```rust
BoxedAggState {                 // 16
    inner: BoxedAggState {      // 16
        return_type: DataType,  // 16
        state: Option<i64>,     // 16
    },
    indices: Vec<usize> [       // 24
        i                       // 8
    ],
}
```

This results in high memory consumption and allocation overhead in `HashAggExecutor`, since there is at least one agg state for each group key.

### Describe the solution you'd like

First, the `indices` vector for projection could be removed. We should project chunks in the executor before feeding them into the aggregator. This method can save 48 bytes.

```rust
BoxedAggState {             // 16
    return_type: DataType,  // 16
    state: Option<i64>,     // 16
}
```

Second, the `return_type` could be omitted as long as it is not composite type (list or struct). This can save another 16 bytes.

```rust
BoxedAggState {             // 16
    state: Option<i64>,     // 16
}
```

Finally, we may further inline common aggregators to completely eliminate allocation in most cases.
```rust
enum AggState {
    Box(BoxedAggState),
    SumI16(Option<i16>),
    SumI32(Option<i32>),
    SumI64(Option<i64>),
    ...
}
```
The state size could be compressed to a limit of 16 bytes.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR decouples aggregate function and its state. This way, function arguments can be moved out of state which reduces state size. resolve #9900 

Currently each aggregate function is defined as a struct that implements `Aggregator` trait. This struct contains both aggregate state and necessary arguments to perform the aggregation (e.g. argument indices, return type and direct arguments). For example:

```rust
struct Sum {
    return_type: DataType,
    // the core state
    state: Option<i64>,
}
impl Aggregator for Sum {
    fn update(&mut self, chunk: &StreamChunk) -> Result<()> {...}
}

// state for each group, takes 96 bytes total
struct Project {
    indices: Vec<usize>,
    inner: Box<dyn Aggregator>,
}
impl Aggregator for Project {
    fn update(&mut self, chunk: &StreamChunk) -> Result<()> {
        let chunk = chunk.project(&self.indices);
        self.inner(&chunk)
    }
}
```

However, in grouped aggregation, different groups have different states, but all of them share the same set of arguments. The current design wastes a lot of memory storing duplicated arguments. 

Inspired by [the design of UDAF API in other systems](https://nightlies.apache.org/flink/flink-docs-release-1.17/docs/dev/table/functions/udfs/#aggregate-functions), this PR introduces the `AggregateFunction` trait and the `AggregateState` struct. The former contains function arguments and has only one instance in the aggregate executor. The latter contains aggregate state and there's one instance per group.

```rust
// state for each group, takes 24 bytes at least
enum AggregateState {
    // for most common aggregates, their state is a single datum.
    Datum(Datum),
    // we also support state of any type
    Any(Box<dyn Any>),
} 

// only one instance
struct Sum {
    return_type: DataType,
}
impl AggregateFunction for Sum {
    fn update(&self, state: &mut AggregateState, chunk: &StreamChunk) -> Result<()> {
        // the function need to downcast the state to a concrete type.
        let state = state.as_datum_mut();
        ...
    }
}

// only one instance
struct Project {
    indices: Vec<usize>,
    inner: Box<dyn AggregateFunction>,
}
impl AggregateFunction for Project {
    fn update(&self, state: &mut AggregateState, chunk: &StreamChunk) -> Result<()> {...}
}
```

Compared with the current way, the new design avoids 2 memory allocations and reduces memory consumption from 96 bytes to 24 bytes for each state.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [x] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)
    This PR significantly reduces the aggregate state size, which may impact the performance of hashagg.

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/expr/src/agg/percentile_cont.rs [hunk 1]
```diff
@@ -64,90 +63,75 @@ use crate::Result;
...
-            data: vec![],
-        }
-    }
+type State = Vec<f64>;
 
-    fn add_datum(&mut self, datum_ref: DatumRef<'_>) {
+impl AggStateDyn for State {}
+
+impl PercentileCont {
+    fn add_datum(&self, state: &mut State, datum_ref: DatumRef<'_>) {
```
Comment (stdrc) [inline src/expr/src/agg/percentile_cont.rs:82]: Ditto

File: src/stream/src/executor/simple_agg.rs [hunk 2]
```diff
@@ -64,6 +64,9 @@ struct ExecutorInner<S: StateStore> {
     /// An operator will support multiple aggregation calls.
     agg_calls: Vec<AggCall>,
```
Comment (stdrc) [inline src/stream/src/executor/simple_agg.rs:65]: Is `agg_calls` still needed after we have `agg_funcs` field?

**Review by stdrc — APPROVED** [review]
LGTM!

File: src/expr/src/agg/mod.rs [hunk 3]
```diff
@@ -37,51 +38,91 @@ mod string_agg;
...
 #[async_trait::async_trait]
-pub trait Aggregator: Send + Sync + DynClone + 'static {
+pub trait AggregateFunction: Send + Sync + 'static {
+    /// Returns the return type of the aggregate function.
     fn return_type(&self) -> DataType;
 
-    /// Update the aggregator with multiple rows.
-    async fn update(&mut self, input: &StreamChunk) -> Result<()>;
+    /// Returns the initial state of the aggregate function.
+    fn init_state(&self) -> AggregateState {
```
[the remark this answers is not in the document]
Comment (wangrunji0408) [reply src/expr/src/agg/mod.rs:48]: Right. `create_state` is better. 😄

File: src/stream/src/executor/simple_agg.rs (hunk 2 above)
Comment (wangrunji0408) [reply src/stream/src/executor/simple_agg.rs:65]: In fact I'm working on a subsequent refactor to merge `agg_calls`, `agg_storages`, `agg_funcs` and other variables into one field, which is something equivalent of `AggregateFunction` in executors.

File: src/expr/src/agg/mod.rs [hunk 4]
```diff
@@ -37,51 +38,91 @@ mod string_agg;
...
+        AggregateState::Datum(None)
+    }
 
-    /// Update the aggregator with a range of rows.
-    async fn update_range(&mut self, input: &StreamChunk, range: Range<usize>) -> Result<()>;
+    /// Update the state with multiple rows.
+    async fn update(&self, state: &mut AggregateState, input: &StreamChunk) -> Result<()>;
```
[the remark this answers is not in the document]
Comment (wangrunji0408) [reply src/expr/src/agg/mod.rs:53]: 🤔 I have no strong option for this.

This method is a combination of `accumulate` and `retract` in functionality. All these names look not so good to me. 😇

File: src/expr/src/agg/percentile_cont.rs (hunk 1 above)
Comment (wangrunji0408) [reply src/expr/src/agg/percentile_cont.rs:82]: This method depends on the direct argument in the function. So I'd like to keep it as it is.

**Pull request merged**